### PR TITLE
Adding storage components to Pyomo dispatcher

### LINKED ISSUE #30: [TASK] Storage in pyomo_solver
--------
Issue Description
--------

Storage constraint writing options should be added to the `pyomo_solver`, and a convention for tracking storage
activity/level should be considered.

Previously, storage dispatch has only been simulated using custom dispatch strategies.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by PaulTalbot-INL**
--------
Pull Request Description
--------
##### What issue does this change request address?
Closes #30

##### What are the significant changes in functionality due to this change request?
Implements storage handling in the pyomo dispatcher, and tests it analytically.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/HERON/wiki/Code-Standards) for details.
- [x] 4. Automated Tests should pass.
- [x] 5. If significant functionality is added, there must be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large tes.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added, the the analytic documentation must be updated/added.
- [x] 9. If any test used as a basis for documentation examples have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by PaulTalbot-INL** [comment]
FYI, while civet is up, not all of our testing machines have firewall exceptions established yet, so the automated testing isn't online yet for HERON as of this writing.

File: tests/integration_tests/storage/transfers.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+
+# Copyright 2020, Battelle Energy Alliance, LLC
+# ALL RIGHTS RESERVED
+"""
+  Implements transfer functions
+"""
+
+def electric_consume(data, meta):
```
Comment (dylanjm) [inline tests/integration_tests/storage/transfers.py:8]: Do these functions need docstrings? This is not a generated python script correct?

File: src/dispatch/pyomo_dispatch.py [hunk 2]
```diff
@@ -446,19 +449,37 @@ def _cashflow_rule(self, meta, m):
...
+          # Production rate for storage defined as R_k = (L_{k+1} - L_k) / dt
+          if t > 0:
+            previous = var[r, t-1]
+            dt = m.Times[t] - m.Times[t-1]
+          else:
+            # FIXME check this with a variety of ValuedParams
+            previous = comp.get_interaction().get_initial_level(meta, None, None, None)
+            dt = m.Times[1] - m.Times[0]
+          new = var[r, t]
+          production = -1 * (new - previous) / dt # swap sign b/c negative is absorbing, positive is emitting
```
Comment (dylanjm) [inline src/dispatch/pyomo_dispatch.py:478]: To be clear, this comment is stating that a negative production value indicates that storage is currently _storing_ energy while positive production is _draining_ it's stored energy?

File: doc/user_manual/src/components_intro.tex [hunk 3]
```diff
@@ -1,3 +1,100 @@
...
+quantity (red, right). That is, the storage unit filly slowly, then fills quickly, then totally depletes before
+filling again. This same activity can be represented by the production rates in the dashed green
+lines, with negative production indicating absorbing into storage and positive production indicating
+emission from storage.
+\begin{figure}[h!]
+  \centering
+  \includegraphics[width=0.5\textwidth]{../pics/storage_activity_explained.png}
+  \caption{Storage Example}
+  \label{fig:storage_example}
+\end{figure}
```
Comment (dylanjm) [inline doc/user_manual/src/components_intro.tex:100]: I think this is a great explanation of interpreting the Storage component.

**Review by dylanjm — DISMISSED** [review]
Looks great, I'll run the tests next and make sure the documentation compiles nicely.

File: src/dispatch/pyomo_dispatch.py (hunk 2 above)
Comment (PaulTalbot-INL) [reply src/dispatch/pyomo_dispatch.py:478]: Yeah, you got it. I'm thinking of just adding a "Conventions" page to the user manual that explains positive/negative signs generally, and moving (or linking) the storage explanation there.

File: tests/integration_tests/storage/transfers.py (hunk 1 above)
Comment (PaulTalbot-INL) [reply tests/integration_tests/storage/transfers.py:8]: Yes, adding those.

**Review by dylanjm — DISMISSED** [review]

**Comment by PaulTalbot-INL** [comment]
It looks like the tests are failing due to a fun little feature, where 5.7 in Conda language means "anything between 5.7 and 5.8", but in the python version checking 5.7 means 5.7.0. ~~I'm not sure what to do about this yet.~~

See fix in https://github.com/idaholab/raven/pull/1327

**Comment by PaulTalbot-INL** [comment]
@dylanjm It looks like the tests are passing; this is ready for review of the new changes.

**Review by dylanjm — APPROVED** [review]

**Pull request merged**